src/util/side_metadata/global.rs

```diff
@@ -72,129 +77,252 @@ impl MappingState {
 ///
 /// * `size` - The size of the source data (in bytes).
 ///
-/// * `global_per_chunk` - The number of bytes of global side metadata required per chunk.
+/// * `global_metadata_spec_vec` - A vector of SideMetadataSpec objects containing all global side metadata.
 ///
-/// * `local_per_chunk` - The number of bytes of policy-specific side metadata required per chunk.
+/// * `local_metadata_spec_vec` - A vector of SideMetadataSpec objects containing all local side metadata.
 ///
 pub fn try_map_metadata_space(
     start: Address,
     size: usize,
-    global_per_chunk: usize,
-    local_per_chunk: usize,
+    global_metadata_spec_vec: &[SideMetadataSpec],
+    local_metadata_spec_vec: &[SideMetadataSpec],
 ) -> bool {
-    let mut aligned_start = start.align_down(BYTES_IN_CHUNK);
-    let aligned_end = (start + size).align_up(BYTES_IN_CHUNK);
-
-    // first chunk is special, as it might already be mapped, so it shouldn't be unmapped on failure
-    let mut munmap_first_chunk: Option<bool> = None;
+    debug_assert!(start.is_aligned_to(BYTES_IN_PAGE));
+    debug_assert!(size % BYTES_IN_PAGE == 0);
+
+    for spec in global_metadata_spec_vec {
+        // nearest page-aligned starting address
```

Comment:
The mapping code for 64bit seems repeating a lot (L94-108, L115-123, L148-157, L164-173, and also in the unmap code) - maybe worth extracing it to a separate function.

Reply:
Done